design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,410 @@
+Think of the following scenario where we have the Ceph and Minio backends both to a level of maturity that we have deemed `stable`.
+Then, if a new storage backend is incorporated into Rook with an early implementation, it should be able to start at `alpha` to properly indicate its nascent level of maturity and stability.
+Therefore, there is a need for each backend to be able to be versioned **independently** of other backends.
+
+Independent versioning is possible by making each storage backend a top level custom resource.
+The same can not be said for an approach of a single cluster CRD that has specific backends as sub-fields.
+To demonstrate this point, let's explore further the scenario above where we have a stable `cluster` CRD that supports `ceph` and `minio`.
+This stable CRD has dropped some properties and gained some properties during its maturity from `alpha` to `stable`.
+To add the support for the brand new storage backend `foostore`, it should be added to the `alpha` CRD because that properly represents its maturity.
+Therefore, we would have the following yaml:
+
+```yaml
+apiVersion: rook.io/v1
+kind: StorageCluster
+spec:
+  someStableField: "stableValue"
+  ceph:
+  mino:
+---
+apiVersion: rook.io/v1alpha1
+kind: StorageCluster
+spec:
+  someAlphaField: "alphaValue"
+  ceph:
+  minio:
+  foostore:
+```
+
+This results in an unfortunate situation for the new `foostore` backend.
+Because its version is coupled with that of the top level cluster CRD, and it is of `alpha` maturity, it is forced to be included in the old `alpha` cluster type, meaning it can't use the latest mature and stable fields.
+This feels like a very negative restriction and suggests that we should decouple storage backends from common cluster abstractions.
+
+Commonality and sharing of types and implementations is important and is another way Rook provides value to the greater storage community beyond the operators that it implements.
+To facilitate this, the following API groups are proposed:
+
+* `rook.io`: common abstractions that have use across multiple storage backends and types
+* `storage.rook.io`: specific storage backend types, such as Ceph.
+
+This would give us the following example manifests that a user might deploy, note how each backend has its own independent version:
+
+```yaml
+apiVersion: storage.rook.io/v1
+kind: Ceph
+spec:
+  cluster: # common cluster scope, properties, etc.
+  monCount: 3
+---
+apiVersion: storage.rook.io/v1beta1
+kind: Minio
+spec:
+  cluster: # common cluster scope, properties, etc.
+  mode: distributed
+---
+apiVersion: storage.rook.io/v1apha2
+kind: Foostore
+spec:
+  cluster: # common cluster scope, properties, etc.
+  fooField: "fooValue"
+```
+
+The `cluster` field used in the preceding example can be for **any** version of the common cluster abstraction that the storage backend wants to reference.
```

Comment:
I think differentiating between common types and backend specific types makes sense for versioning reasons. However, syntactically I would hope we can follow the lead of `compositecontroller` in the meta-controller project to make the YAML experience more natural. Specifically,

```
StorageCluster (independent CRD)
     common fields
     ceph (independent CRD)
          fields
     minio (independent CRD)
```

Is something like that possible? I wonder if @enisoc could weigh in here relative to metacontroller.

Reply:
Some more research needs to be done here, but I'll get to the bottom of this and reach out to Anthony as needed.

Reply:
In my opinion, the choice of whether to have `StorageCluster` or (for example) `Ceph` as the top-level resource should come down to how the controllers work (which I don't know). If there will be a single controller that performs abstract tasks by calling into various backends, then `StorageCluster` should be a top-level resource with fields named `ceph`, `minio`, etc. If Ceph is served by its own controller, and Minio by a totally different controller, then `Ceph` and `Minio` should be distinct, top-level resource kinds.

From the rest of this doc, it sounds like `StorageCluster` is not going to need its own controller. In that sense, it's more like the `ObjectMeta` type in `meta/v1` -- a versioned, embedded struct type that is not actually a resource of its own. If that's the case, I like the proposed design here where `Ceph` is top-level and embeds `StorageCluster` to share struct types between backends.